rewrite-core/src/main/java/org/openrewrite/trait/Trait.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.trait;
+
+import org.openrewrite.Cursor;
+import org.openrewrite.Incubating;
+import org.openrewrite.Tree;
+
+/**
+ * A trait captures semantic information related to a syntax element.
+ *
+ * @param <T> The type of the tree that this trait is related to. When
+ *            multiple specific types of tree are possible, this should
+ *            be the lowest common super-type of all the types.
```

Comment:
I was about to raise a concern about multiple specific types of trees, but I think you addressed it in the second sentence.

Reply:
Yeah, I have more thinking on this, and the most obvious case I know of right now is a trait for `JavaDependency` which could return a Gradle `J.MethodInvocation` or an Maven (XML) `Xml.Tag` which have `Tree` as a lowest common super-type. Similar for the trait of `SqlQuery`. I have a plan, just need to write it out.